### PR TITLE
Add process for building cellranger index to build-index.nf

### LINKED ISSUE #65: Include cellranger index in build-index.nf
Right now we only include the indices for Alevin-fry in `build-index.nf`, however, we will be using Spaceranger for spatial transcriptomics libraries and will need to build a separate index for that. In doing this we should make sure that the ensembl versions that we are building from is consistent across all of our indices that we are using. We should use a similar setup to how we were previously building the [cellranger index in `alsf-scpca`](https://github.com/AlexsLemonade/alsf-scpca/blob/main/workflows/rnaseq-ref-index/build-cellranger-index.nf).

### PR DESCRIPTION
**Pull request by allyhawkins**
Closes #65. This PR adds the `cellranger_index` process to the workflow for building the indices, `build-index.nf` ensuring that when we build all of our indices they will use the same assembly. I followed the same setup that was previously used in generating the cellranger index in [`alsf-scpca/workflows/rnaseq-ref-index/build-cellranger-index.nf`](https://github.com/AlexsLemonade/alsf-scpca/blob/544673eb008843548ac5c8663da65d4732d1f158/workflows/rnaseq-ref-index/build-cellranger-index.nf). 

This was pretty straight forward and I was able to test it and everything ran successfully, producing a new index with ensembl 104. The only question I had was about how we wanted to name this index. I included the assembly name that we use in the splici index and then appended `_cdna_cellranger`, but let me know if there is a different naming convention that would be preferred.

### CODE REVIEW

File: build-index.nf [hunk 1]
```diff
@@ -66,10 +66,41 @@ process salmon_index{
...
+    path(cellranger_index)
+  script:
+    cellranger_index = "${assembly}_cdna_cellranger"
+    """
+    gunzip -c ${fasta} > genome.fasta
+    gunzip -c ${gtf} > genome.gtf
+    cellranger mkgtf \
+      genome.gtf \
+      filtered.gtf \
+      --attribute=gene_biotype:protein_coding
```
Comment (jashapiro) [inline build-index.nf:87]: Is this going to be too stringent? I would think we should include ncRNA and such: The default cellranger index includes much more: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/references#mkgtf

(I just looked at the full list of `gene_biotypes` in the reference we are using, which appears below (generated with `grep -o 'gene_biotype [^;]*' Homo_sapiens.GRCh38.104.gtf |sort | uniq`) and noted that the code used by 10X might miss `lncRNA` as it is looking for `lincRNA`, according to that page.) 

Alternatively, we could probably just not filter the gtf file? This would probably be more comparable to what we have for the salmon index, which I do not think we filter at all (hence 60K "genes").


```
gene_biotype "IG_C_gene"
gene_biotype "IG_C_pseudogene"
gene_biotype "IG_D_gene"
gene_biotype "IG_J_gene"
gene_biotype "IG_J_pseudogene"
gene_biotype "IG_V_gene"
gene_biotype "IG_V_pseudogene"
gene_biotype "IG_pseudogene"
gene_biotype "Mt_rRNA"
gene_biotype "Mt_tRNA"
gene_biotype "TEC"
gene_biotype "TR_C_gene"
gene_biotype "TR_D_gene"
gene_biotype "TR_J_gene"
gene_biotype "TR_J_pseudogene"
gene_biotype "TR_V_gene"
gene_biotype "TR_V_pseudogene"
gene_biotype "lncRNA"
gene_biotype "miRNA"
gene_biotype "misc_RNA"
gene_biotype "polymorphic_pseudogene"
gene_biotype "processed_pseudogene"
gene_biotype "protein_coding"
gene_biotype "pseudogene"
gene_biotype "rRNA"
gene_biotype "rRNA_pseudogene"
gene_biotype "ribozyme"
gene_biotype "sRNA"
gene_biotype "scRNA"
gene_biotype "scaRNA"
gene_biotype "snRNA"
gene_biotype "snoRNA"
gene_biotype "transcribed_processed_pseudogene"
gene_biotype "transcribed_unitary_pseudogene"
gene_biotype "transcribed_unprocessed_pseudogene"
gene_biotype "translated_processed_pseudogene"
gene_biotype "translated_unprocessed_pseudogene"
gene_biotype "unitary_pseudogene"
gene_biotype "unprocessed_pseudogene"
gene_biotype "vault_RNA"
```

File: build-index.nf [hunk 2]
```diff
@@ -66,10 +66,41 @@ process salmon_index{
...
+  publishDir "${params.ref_dir}/cellranger_index", mode: 'copy'
+  label 'cpus_8'
+  input:
+    path(fasta)
+    path(gtf)
+    val(assembly)
+  output:
+    path(cellranger_index)
+  script:
+    cellranger_index = "${assembly}_cdna_cellranger"
```
Comment (jashapiro) [inline build-index.nf:80]: If we are not filtering, I would probably shorten this to:
```suggestion
    cellranger_index = "${assembly}_cellranger"
```
or perhaps 
```suggestion
    cellranger_index = "${assembly}_cellranger_full"
```

Either way, this should probably be added to the `nextflow.config` `params` block.

**Review by jashapiro — COMMENTED** [review]
This looks good. I made a couple of suggestions below, which might make this index more comparable to the one we use for salmon.

I was going to make an additional suggestion about naming params, but ended up putting it in a separate issue: https://github.com/AlexsLemonade/scpca-nf/issues/67

File: build-index.nf (hunk 1 above)
Comment (allyhawkins) [reply build-index.nf:87]: We definitely don't do any filtering like this for the salmon index, so I can remove this to make it more comparable.

**Comment by allyhawkins** [comment]
I went ahead and removed filtering and it worked fine, creating a larger index this time.

File: build-index.nf [hunk 3]
```diff
@@ -66,10 +66,40 @@ process salmon_index{
...
+      --nthreads=${task.cpus}
+    """
+}
+
+
 
 workflow {
   // generate splici and spliced cDNA reference fasta
-  generate_reference(params.gtf, params.fasta, params.assembly)
+  //generate_reference(params.gtf, params.fasta, params.assembly)
```
Comment (jashapiro) [inline build-index.nf:100]: Uncommenting lines we want to keep
```suggestion
  generate_reference(params.gtf, params.fasta, params.assembly)
```

File: build-index.nf [hunk 4]
```diff
@@ -66,10 +66,40 @@ process salmon_index{
...
   // create index using reference fastas
-  salmon_index(generate_reference.out.fasta_files, params.fasta)
+  //salmon_index(generate_reference.out.fasta_files, params.fasta)
```
Comment (jashapiro) [inline build-index.nf:102]: ditto
```suggestion
  salmon_index(generate_reference.out.fasta_files, params.fasta)
```

**Review by jashapiro — COMMENTED** [review]
Looks good, with two small suggestions I made, and one that I can't actually put in, which is to add 
```
  cellranger_index = "s3://nextflow-ccdl-data/reference/homo_sapiens/ensembl-104/cellranger_index/Homo_sapiens.GRCh38.104_cellranger_full"
```
to the params in `nextflow.config`

File: build-index.nf [hunk 5]
```diff
@@ -66,10 +66,40 @@ process salmon_index{
...
+    gunzip -c ${fasta} > genome.fasta
+    gunzip -c ${gtf} > genome.gtf
+    cellranger mkgtf \
+      genome.gtf \
+      filtered.gtf
+    
+    cellranger mkref \
+      --genome=${cellranger_index} \
+      --fasta=genome.fasta \
+      --genes=filtered.gtf \
```
Comment (jashapiro) [inline build-index.nf:91]: I think you can actually completely skip `mkgtf` as I don't think it is doing anything here.
```suggestion
    cellranger mkref \
      --genome=${cellranger_index} \
      --fasta=genome.fasta \
      --genes=genome.gtf \
```

**Comment by allyhawkins** [comment]
@jashapiro I went ahead and made the edits that you suggested and then also had to increase the memory that the process is using. I noticed that it was only making some of the index files but not all of them even though it was completing the process successfully without error. It wasn't until I attempted to run spaceranger using the index later on that I realized there was missing files and only a partial index so had to bump up the memory required too.

**Review by jashapiro — APPROVED** [review]
LGTM!
